redis/src/push_manager.rs

```diff
@@ -0,0 +1,51 @@
+use crate::{PushKind, Value};
+use std::sync::mpsc::SyncSender;
+
+/// Holds information about received Push data
+#[derive(Debug, Clone)]
+pub struct PushInfo {
+    /// Push Kind
+    pub kind: PushKind,
+    /// Data from push message
+    pub data: Vec<Value>,
+}
+/// `PushSenderType` holds multiple types of mpsc channels
+#[derive(Clone)]
+pub enum PushSenderType {
+    /// Tokio mpsc Sender
+    #[cfg(feature = "tokio-comp")]
+    Tokio(tokio::sync::mpsc::Sender<PushInfo>),
+    /// Standard mpsc SyncSender
+    Standard(SyncSender<PushInfo>),
+}
+
+/// Manages Push messages for both tokio and std channels
+#[derive(Clone, Default)]
+pub struct PushManager {
+    sender: Option<PushSenderType>,
+}
+impl PushManager {
+    /// Try to send `PushInfo` to mpsc channel without blocking
+    /// if the function returns false it means
+    /// `PushManager` has no channel or provided channel is full.
+    pub fn send(&self, pi: PushInfo) -> bool {
+        if let Some(sender) = &self.sender {
+            return match sender {
+                #[cfg(feature = "tokio-comp")]
+                PushSenderType::Tokio(tokio_sender) => tokio_sender.try_send(pi).is_ok(),
+                PushSenderType::Standard(std_sender) => std_sender.try_send(pi).is_ok(),
+            };
+        }
+        false
+    }
+
+    /// Checks if `PushManager` has provided any channel.
+    pub fn has_sender(&self) -> bool {
```

Comment:
since `send` doesn't do anything if there's no sender, why is this check needed?

Reply:
the reason is sometimes we have to clone values to create a PushInfo, we can avoid it with checking if have sender or not.

Reply:
why are the values cloned? Can't they be moved? Is someone accessing them after sending?

Reply:
read_response returns Value, so function has to clone it for sending into channel.